[BUG] How to specify TENSORFLOW_ROOT during install?

### Bug summary

I'm trying to install deempd-kit on a HPC cluster from source with GPU support with a custom prefix. Tensorflow is already installed, but during the installation, pip fails to find the tensorflow headers (which are there).

Install command:
pip install --prefix=$PREFIX -vvv deepmd-kit/ --no-build-isolation --no-dependencies

Importing tensorflow in python works fine and executing `python -c "import tensorflow as tf;print(tf.reduce_sum(tf.random.normal([1000, 1000])))"` also works.
Just skbuild can't find the headers, which are here for example:
`/p/software/juwelsbooster/stages/2022/software/TensorFlow/2.6.0-gcccoremkl-11.2.0-2021.4.0-CUDA-11.5/lib/python3.9/site-packages/tensorflow/include/tensorflow/core/public/session.h`

How do I tell the installation where to find those headers?


### DeePMD-kit Version

2.1.4

### TensorFlow Version

2.6.0

### How did you download the software?

Built from source

### Input Files, Running Commands, Error Log, etc.

Relevant output:
```
  -- Configuring incomplete, errors occurred!
  See also "/tmp/pip-req-build-u5kc1d1a/_cmake_test_compile/build/CMakeFiles/CMakeOutput.log".
  Not searching for unused variables given on the command line.
  -- The C compiler identification is GNU 11.2.0
  -- Detecting C compiler ABI info
  -- Detecting C compiler ABI info - done
  -- Check for working C compiler: /p/software/juwelsbooster/stages/2022/software/GCCcore/11.2.0/bin/cc - skipped
  -- Detecting C compile features
  -- Detecting C compile features - done
  -- The CXX compiler identification is GNU 11.2.0
  -- Detecting CXX compiler ABI info
  -- Detecting CXX compiler ABI info - done
  -- Check for working CXX compiler: /p/software/juwelsbooster/stages/2022/software/GCCcore/11.2.0/bin/c++ - skipped
  -- Detecting CXX compile features
  -- Detecting CXX compile features - done
  -- Configuring done
  -- Generating done
  -- Build files have been written to: /tmp/pip-req-build-u5kc1d1a/_cmake_test_compile/build
  -- The C compiler identification is GNU 11.2.0
  -- The CXX compiler identification is GNU 11.2.0
  -- Detecting C compiler ABI info
  -- Detecting C compiler ABI info - done
  -- Check for working C compiler: /p/software/juwelsbooster/stages/2022/software/GCCcore/11.2.0/bin/cc - skipped
  -- Detecting C compile features
  -- Detecting C compile features - done
  -- Detecting CXX compiler ABI info
  -- Detecting CXX compiler ABI info - done
  -- Check for working CXX compiler: /p/software/juwelsbooster/stages/2022/software/GCCcore/11.2.0/bin/c++ - skipped
  -- Detecting CXX compile features
  -- Detecting CXX compile features - done
  -- Found Git: /usr/bin/git (found version "2.31.1")
  -- Supported model version: 1.1
  -- Found CUDA: /p/software/juwelsbooster/stages/2022/software/CUDA/11.5 (found version "11.5")
  -- Found CUDA in /p/software/juwelsbooster/stages/2022/software/CUDA/11.5, build nv GPU support
  -- Will not build AMD GPU support
  CMake Error at cmake/Findtensorflow.cmake:84 (message):
    Not found 'include/tensorflow/core/public/session.h' directory or other
    header files in path
    '/p/software/juwelsbooster/stages/2022/software/TensorFlow/2.6.0-gcccoremkl-11.2.0-2021.4.0-CUDA-11.5/lib/python3.9/site-packages/tensorflow;/p/software/juwelsbooster/stages/2022/software/TensorFlow/2.6.0-gcccoremkl-11.2.0-2021.4.0-CUDA-11.5/lib/python3.9/site-packages/tensorflow/../tensorflow_core;/p/software/juwelsbooster/stages/2022/software/TensorFlow/2.6.0-gcccoremkl-11.2.0-2021.4.0-CUDA-11.5/lib/python3.9/site-packages/tensorflow/../../../..'
    You can manually set the tensorflow install path by -DTENSORFLOW_ROOT
  Call Stack (most recent call first):
    CMakeLists.txt:88 (find_package)


  -- Configuring incomplete, errors occurred!
  See also "/tmp/pip-req-build-u5kc1d1a/_skbuild/linux-x86_64-3.9/cmake-build/CMakeFiles/CMakeOutput.log".
    File "/p/project/eat2d/libs/install_deepmd_gpu/lib/python3.9/site-packages/skbuild/setuptools_wrap.py", line 637, in setup
      env = cmkr.configure(
    File "/p/project/eat2d/libs/install_deepmd_gpu/lib/python3.9/site-packages/skbuild/cmaker.py", line 328, in configure
      raise SKBuildError(
```

### Steps to Reproduce

Log in to HPC, load preinstalled python and tensorflow module, try to install DeepMD with pip with custom prefix.

### Further Information, Files, and Links

_No response_

--- a/setup.py
+++ b/setup.py
@@ -85,13 +85,13 @@ except (AttributeError, TypeError, IndexError):
     # setuptools will re-find tensorflow after installing setup_requires
     tf_install_dir = None
 
-# add cmake as a build requirement if cmake>3.12 is not installed
+# add cmake as a build requirement if cmake>=3.16 is not installed
 try:
     cmake_version = get_cmake_version()
 except SKBuildError:
     setup_requires.append("cmake")
 else:
-    if cmake_version in SpecifierSet("<3.12"):
+    if cmake_version in SpecifierSet("<3.16"):
         setup_requires.append("cmake")
 
 Path("deepmd").mkdir(exist_ok=True)

--- a/source/CMakeLists.txt
+++ b/source/CMakeLists.txt
@@ -1,4 +1,5 @@
-cmake_minimum_required(VERSION 3.12)
+# new in 3.16: GET_RUNTIME_DEPENDENCIES
+cmake_minimum_required(VERSION 3.16)
 project(DeePMD)
 
 option(BUILD_TESTING "Enable converage" OFF)

--- a/source/cmake/Findtensorflow.cmake
+++ b/source/cmake/Findtensorflow.cmake
@@ -56,7 +56,6 @@ if(BUILD_PY_IF)
 	# here TENSORFLOW_ROOT is path to site-packages/tensorflow
 	# for conda libraries, append extra paths
 	list(APPEND TensorFlow_search_PATHS "${TENSORFLOW_ROOT}/../tensorflow_core")
-	list(APPEND TensorFlow_search_PATHS "${TENSORFLOW_ROOT}/../../../..")
 endif()
 
 # includes
@@ -71,14 +70,6 @@ find_path(TensorFlow_INCLUDE_DIRS
   PATH_SUFFIXES "/include"
   NO_DEFAULT_PATH
   )
-find_path(TensorFlow_INCLUDE_DIRS_GOOGLE
-  NAMES 
-  google/protobuf/type.pb.h
-  PATHS ${TensorFlow_search_PATHS} 
-  PATH_SUFFIXES "/include"
-  NO_DEFAULT_PATH
-  )
-list(APPEND TensorFlow_INCLUDE_DIRS ${TensorFlow_INCLUDE_DIRS_GOOGLE})
   
 if (NOT TensorFlow_INCLUDE_DIRS AND tensorflow_FIND_REQUIRED)
   message(FATAL_ERROR 
@@ -149,6 +140,44 @@ foreach (module ${TensorFlowFramework_FIND_COMPONENTS})
       "You can manually set the tensorflow install path by -DTENSORFLOW_ROOT ")
   endif ()
 endforeach ()
+
+# find protobuf header
+find_path(TensorFlow_INCLUDE_DIRS_GOOGLE
+  NAMES 
+  google/protobuf/type.pb.h
+  PATHS ${TensorFlow_search_PATHS} 
+  PATH_SUFFIXES "/include"
+  NO_DEFAULT_PATH
+  )
+message(STATUS proto ${TensorFlow_INCLUDE_DIRS_GOOGLE})
+if (NOT TensorFlow_INCLUDE_DIRS_GOOGLE)
+  message(STATUS "Protobuf headers are not found in the directory of TensorFlow, assuming external protobuf was used to build TensorFlow")
+  # try to find from ldd list of TF library
+  # a warning is threw here, just ignore it
+  file(GET_RUNTIME_DEPENDENCIES
+    RESOLVED_DEPENDENCIES_VAR TensorFlow_LINKED_LIBRARIES
+    LIBRARIES ${TensorFlowFramework_LIBRARY}
+    POST_INCLUDE_REGEXES "^.+protobuf\..+$"
+  )
+  # search protobuf from linked libraries
+  foreach(_lib ${TensorFlow_LINKED_LIBRARIES})
+      string(REGEX MATCH "^.+protobuf\..+$" _protobuf_lib ${_lib})
+      if (_protobuf_lib)
+        set(Protobuf_LIBRARY ${_protobuf_lib})
+        break()
+      endif()
+  endforeach()
+  if (NOT Protobuf_LIBRARY)
+    message(FATAL_ERROR "TensorFlow is not linked to protobuf")
+  endif()
+  get_filename_component(Protobuf_LIBRARY_DIRECTORY ${Protobuf_LIBRARY} DIRECTORY)
+  # assume the include directory is ../include
+  set(Protobuf_INCLUDE_DIR ${Protobuf_LIBRARY_DIRECTORY}/../include)
+  find_package(Protobuf REQUIRED)
+  message(STATUS ${Protobuf_INCLUDE_DIR})
+  set(TensorFlow_INCLUDE_DIRS_GOOGLE ${Protobuf_INCLUDE_DIRS})
+endif()
+list(APPEND TensorFlow_INCLUDE_DIRS ${TensorFlow_INCLUDE_DIRS_GOOGLE})
 
 if (BUILD_CPP_IF)
   # define the output variable

--- a/source/cmake/Findtensorflow.cmake
+++ b/source/cmake/Findtensorflow.cmake
@@ -174,7 +174,6 @@ if (NOT TensorFlow_INCLUDE_DIRS_GOOGLE)
   # assume the include directory is ../include
   set(Protobuf_INCLUDE_DIR ${Protobuf_LIBRARY_DIRECTORY}/../include)
   find_package(Protobuf REQUIRED)
-  message(STATUS ${Protobuf_INCLUDE_DIR})
   set(TensorFlow_INCLUDE_DIRS_GOOGLE ${Protobuf_INCLUDE_DIRS})
 endif()
 list(APPEND TensorFlow_INCLUDE_DIRS ${TensorFlow_INCLUDE_DIRS_GOOGLE})

